flux-cron: cron-like service

grondo opened the pull request:
This PR contains a somewhat simplified cron-like service for a flux instance.

It isn't quite ready (no tests and documentation), but I thought I'd open up the PR now for feedback, as there may be a lot.

TODO
- [x]  Tests
- [x]  Docs
- [x]  Fix cron entry counter reset on `flux cron start`
- [x]  Add entry repeat count to `flux cron list` default output
- [x]  Sync cron_entry_run() to `hb` event by default for all entries
- [x]  Fix `-r, --rank` option
- [x]  Add entry common options for `flux cron event ...`
- [x]  Load cron module in default rc1 script?
- [x]  cron task timeout support

## cron module

The cron module itself has the following components:
 - A `cron_task` type, which is an asynchronous exec abstraction built around `cmb.exec` service
 - A `cron_entry` type, which contains the data you'd expect for cron entries: unique id, name, command to run, current and historical `cron_task` information, etc
 - An "entry type" class which controls when the cron entry is executed

Current types implemented in this PR thus far are:
 * Interval: Run periodically using a `flux_timer_watcher`
 * Event: Run once per event, or optionally once every Nth event

Other types can be easily added, though all I can think of is to add a real cron-like "date&time" type built on the as-yet-unported-to-flux-handle libev [ev_periodic](http://pod.tst.eu/http://cvs.schmorp.de/libev/ev.pod#code_ev_periodic_code_to_cron_or_not) watcher.

Cron entries by default are run indefinitely until stopped or deleted,  but a `repeat` flag is supported which can allow an `at(1)` like service as well

## flux-cron utility

Included is a front-end `flux-cron` utility which allows the creation/deletion, start/stop, and query of flux-cron jobs. A few examples will probably sum it up nicely:
```
grondo@flux-core:~ $ flux cron help
Usage: flux-cron [OPTIONS] COMMAND [OPTIONS...]
  -h, --help                 Display this message.
Supported commands:
 delete       Delete flux cron entries
 help         Display this help or help for a command
 list         List registered and stopped fluxicron jobs
 interval     run COMMAND once every INTERVAL
 start        Start a stopped flux cron job
 event        run COMMAND on every event matching TOPIC
 stop         Stop a flux cron job from executing
```

Create a job to run every 10s:
```
grondo@flux-core:~ $ flux cron interval 10s date
interval: cron-2 created
grondo@flux-core:~ $ flux cron list
    ID CMD/NAME        STATE      #RUNS            LASTRUN             STATUS
     2 date            Active         0               None               None
grondo@flux-core:~ $ flux cron list
    ID CMD/NAME        STATE      #RUNS            LASTRUN             STATUS
     2 date            Active         1      8 seconds ago             Exited
```
```
grondo@flux-core:~ $ flux cron interval --help
Usage: flux-cron interval INTERVAL COMMAND
  -h, --help                 Display this message.
Options:
  -R, --repeat=N             Run COMMAND at most N times
  -c, --history-count=N      Save information on past N runs in cron history
  -r, --rank=RANK            target RANK for COMMAND execution (Default = 0)
  -N, --name=S               Name cron job string S instead of default
```


In one second run `false`, but only run it once:
Note non-zero exit code shows status as `FAILED`.
```
grondo@flux-core:~ $ flux cron interval --repeat=1 1s false
interval: cron-3 created
grondo@flux-core:~ $ flux cron list
    ID CMD/NAME        STATE      #RUNS            LASTRUN             STATUS
     2 date            Active         5      2 seconds ago             Exited
     3 false           Stopped        1      3 seconds ago             Failed
```

Start/stop executing jobs:
```
grondo@flux-core:~ $ flux cron stop 2
Stopped cron-2
grondo@flux-core:~ $ flux cron list
    ID CMD/NAME        STATE      #RUNS            LASTRUN             STATUS
     2 date            Stopped        6     46 seconds ago             Exited
     3 false           Stopped        1     57 seconds ago             Failed
grondo@flux-core:~ $ flux cron start 2
Started cron-2
grondo@flux-core:~ $ flux cron list
    ID CMD/NAME        STATE      #RUNS            LASTRUN             STATUS
     2 date            Active         6       a minute ago             Exited
     3 false           Stopped        1       a minute ago             Failed
```

flux-cron entries stay resident until they are deleted with `flux cron delete`
```
grondo@flux-core:~ $ flux cron delete 2
Removed cron-2: date last ran a second ago for 24.03ms
```

grondo commented:
Don't pay too much attention to tests just yet, but I've been periodically seeing this failure from Travis during development from the `popen2` tests:
```
not ok 14 - write to fd worked
#   Failed test 'write to fd worked'
#   at test/popen2.c line 59.
```

garlick commented:
This looks really great @grondo.  I'll play around with it a bit.

First I'll have a look at that popen2 test failure.   It's probably racy...

garlick commented:
Instrumented that test and found the write call is failing with EPIPE.  Looking a bit closer, it's racy and doesn't really add much, so I've deleted in my pr #658.  Feel free to cherry pick 9da305e104f7e0c089b8dba3c253612797041508 and which ever PR goes in after the other can drop it later.

grondo commented:
Thanks @garlick, Like I said I haven't done testing here yet so apologies if it isn't robust yet.

Other notes:
 - Output from cron jobs is logged via `flux_log`. It would be trivial to optionally save output with the cron_task object so that it is accessible via `flux-cron`, however I didn't want to grow RSS of broker unnecessarily.
 - A new cron_task is not launched at the next interval/event if the current task is still running. This is logged but not recorded any other way yet.
 - cron jobs are always run as `sh -c 'command...'`, but there's no reason that has to be the case (an opion?)
 - I was considering adding a timeout option for cron jobs (also trivial). 
 - Perhaps it makes more sense for the `flux-cron` usage for job creation to look like

   ```
flux cron create [OPTIONS] TYPE ARGS COMMAND
   ```

   Where `[OPTIONS]` are the common cron entry options and `ARGS` are type-specific options.

Also, I was thinking of how a cron job could easily be triggered by an interval *and* an event, e.g. run every 30m but only on the next `hb` event, or run after every 1000 `wrexec.state.complete` events but only after 5 minutes have expired, etc... No great ideas yet, perhaps specific cron types for these cases.

Specifically, it would be pretty easy to synchronize `cron_task_run` on hb optionally... maybe that at least should be done?

garlick commented:
> Also, I was thinking of how a cron job could easily be triggered by an interval and an event, e.g. run every 30m but only on the next hb event,

This in particular strikes me as a good idea, in fact could "synchronize on heartbeat" be worked in as a default some way?   For example, requiring an option to _not_ be synchronized on the heartbeat?

Possibly this is as designed but I found it a bit unexpected.  Looks like `--repeat` works for creating the cron job, but is ignored if the job is restarted.
```
$ flux module load cron
$ flux  cron interval --repeat=1 1s flux logger crontest1
interval: cron-1 created
$ flux cron list
    ID CMD/NAME        STATE      #RUNS            LASTRUN             STATUS
     1 flux            Stopped        1      3 seconds ago             Exited
$ flux cron start 1
# pause a few secs
$ flux cron list
    ID CMD/NAME        STATE      #RUNS            LASTRUN             STATUS
     1 flux            Active        37       a second ago             Exited
```

garlick commented:
> Output from cron jobs is logged via flux_log

That seems like a good default to me actually.  I would be fine with deferring any output redirection options till we have a demonstrated need.

> I was considering adding a timeout option for cron jobs (also trivial).

Maybe that would be best deferred to an external program like timeout(1)?

grondo commented:
> Maybe that would be best deferred to an external program like timeout(1)?

I guess that works, but I'm not exactly following why it would be "best". It does require a different command for every timeout you might want, and doesn't allow any kind of default, unless it is ok to modify command line of all entries. Also, while this probably isn't an issue anymore there used to be multiple versions of `timeout(1)` with different argument handling.
Finally, a native timeout would allow the cron job state to display "timedout"

However, I suppose there is an argument for keeping it simple, but I don't think we should mimic *everything* about good 'ol cron...

grondo commented:
Also, while it is supported, cron jobs currently can't have custom environment nor working dir, they just inherit from the broker.

grondo commented:
> Possibly this is as designed but I found it a bit unexpected. Looks like --repeat works for creating the cron job, but is ignored if the job is restarted.

Thanks that is a bug, `start` should reset the counter but keep the value of repeat.

grondo commented:
> in fact could "synchronize on heartbeat" be worked in as a default some way? For example, requiring an option to not be synchronized on the heartbeat?

Yeah, I think that could be done and is a good idea.. I'll try to see how it goes.

garlick commented:
Good points on timeout(1) - it seems like a native timeout would be a good feature.

>Thanks that is a bug, start should reset the counter but keep the value of repeat.

Should `cron list` show the repeat values then so I know what I'm going to get if I `start` a job listed there?

garlick commented:
> Also, while it is supported, cron jobs currently can't have custom environment nor working dir, they just inherit from the broker.

I'd be tempted to make FLUX_CRON_ID a default but I can't think of a use case so maybe not...

grondo commented:
> Should cron list show the repeat values then so I know what I'm going to get if I start a job listed there?

Good idea, ok to fold it into the #RUNS column? `cron list` is also missing other data like the type of entry, the trigger (timeout or event topic), etc. I was planning to add an `info` or other subcommand to dump detailed information on an ID.

garlick commented:
> ok to fold it into the #RUNS column?
Yeah whatever you think makes sense!  An info command definitely sounds good.

I'm having trouble getting the `-r` option to work. Just to confirm: the cron module would be loaded on rank 0 only, and the rank option would cause it to invoke the exec service on the specified node, correct?

grondo commented:
> I'm having trouble getting the -r option to work. Just to confirm: the cron module would be loaded on rank 0 only, and the rank option would cause it to invoke the exec service on the specified node, correct?

That is the idea, but I haven't even tested it yet so I'm not surprised if it doesn't work. Perhaps I should work on writing tests asap!

grondo commented:
I added a TODO list at the top of the Issue so I don't lose track of some of these items. Feel free to add anything there if you want!

grondo commented:
Ok, just pushed an update 
 
 * Added `t0015-cron.t` sharness tests
 * Fixed entry counter reset on `flux cron start ID`
 * List repeat count in `flux cron list` if set under `#RUNS` as `<count>/<repeat>`
 * Fixed `flux cron [type] --rank` option
 * Changed `-R, --repeat` option to `-c, --count` to match other `flux` utilities
 * Added `flux cron dump ID` command to dump a detailed list of all cron entry keys to stdout
 * Fixed handling of cron tasks that fail to exec or fail cmb.exec

garlick commented:
Nice!  I'll kick the tires tomorrow - have to leave for home and then I'm tied up this evening.

grondo commented:
Just added experimental support for synchronizing all cron task execution to an event. Right now this option is only activated when the `sync=` option is passed to the module, e.g. `sync=hb` should synchronize all tasks with a heartbeat event. 

While tasks are deferred, cron treats them as running or active and refuses to schedule multiple tasks at once (e.g. an interval much smaller than the heartbeat interval won't cause task requests to pile up). Skipped task requests do not count against the repeat count of the cron entry.

This doesn't have any testing yet.

grondo commented:
Hm, this will currently break flux-cron event entries using the same topic as `sync=topic`. They would be forced to defer until the next event. Have to think about how to handle that in general...

grondo commented:
Oops, also noticed there's still some debug printfs in the current code. Sorry about that...

garlick commented:
Maybe the output should include the rank the cron job executed on?
```
$ flux cron interval -c1 -r2 1s echo hello
interval: cron-4 create
$ flux dmesg | grep cron-4
[1462467847.162279] cron.info[0]: cron-4[echo]: command=echo hello: hello
```
Also it appears that `-r all` or ranges like `-r 1-3` are accepted but doesn't have the expected effect?
```
$ flux cron interval -c1 -r all 1s echo xyz
interval: cron-6 created
$ flux dmesg|grep cron-6
[1462468045.351752] cron.info[0]: cron-6[echo]: command=echo xyz: xyz
```

grondo commented:
Ok, great comments. The `flux cron` command should give an error on those `--rank` options as multiple ranks is not supported. Do you think we need to support multi-rank (parallel) execution directly in cron? It won't be impossible, but I was thinking people could use `flux cron <type> <arg> flux exec...` if need be to avoid dealing with parallel execution in cron.

grondo commented:
Ok, the latest push allows cron tasks to be executed up to a small amount of time *after* the synchronization event has fired. This value is 15ms by default (picked randomly!), but can be set on the `module load` command line with a `sync_epsilon=D` parameter, where D is a floating point value in seconds. A value of 0 is allowed and would mean tasks always wait until the next event.

I am not sure how the order of message handlers for the same message is managed, if possible it would be nice to ensure the deferred handler runs first, otherwise a cron entry scheduled on the same event as the sync event may get called first, unnecessarily scheduling its task to the deferred list, which then runs immediately. It doesn't change the timing at all, but is a bit of unecessary work.

grondo commented:
Hm, I seem to have introduced a hang in `module remove`. The cron module is stuck in `flux_event_unsubscribe` (for an event-type cron entry) while the broker is stuck in `module_destroy->pthread_join)

Should I not be unsubscribing from events during module exit? Or, more likely, I introduced some bug I don't see in `cron_ctx_destroy()`.

I'll check to see if this is reproducible.

garlick commented:
> I was thinking people could use flux cron <type> <arg> flux exec... if need be to avoid dealing with parallel execution in cron.

Would it make sense to drop -r and just have people run flux exec in that case too?  I know you almost get that for free, but it feels incomplete to me if it doesn't work like -r in other places.

> Should I not be unsubscribing from events during module exit?

That _should_ be OK.  The `pthread_join()` is not supposed to be called until the module status transitions to `FLUX_MODSTATE_EXITED`, which isn't sent from the module thread until after `mod_main()` exits.  So that state you reported should not happen.  Strange.

grondo commented:
> That should be OK. The pthread_join() is not supposed to be called until the module status transitions to FLUX_MODSTATE_EXITED, which isn't sent from the module thread until after mod_main() exits. So that state you reported should not happen. Strange.

It seems to be reproducible on this branch with the following:
```
$ src/cmd/flux start
(flux-0.0-19) grondo@hype356:~/git/flux-core.git/el6$ flux module load cron
(flux-0.0-19) grondo@hype356:~/git/flux-core.git/el6$ flux cron event foo echo hi
event: cron-1 created
(flux-0.0-19) grondo@hype356:~/git/flux-core.git/el6$ flux module remove cron
(flux-0.0-19) grondo@hype356:~/git/flux-core.git/el6$ flux ping cmb
```

The broker main loop is hung waiting on `pthread_join` so the ping just hangs.

I wonder if this would be fixed by using some other non-blocking method to ensure the module thread has exited, e.g. Linux seems to have `pthread_tryjoin_np(3)`, though I'm not sure you could trigger on an event to retry the join, you could implement a timeout&&kill to ensure the broker loop stays active at least.

I'm still guessing I've done something dumb here...

grondo commented:
> Would it make sense to drop -r and just have people run flux exec in that case too? I know you almost get that for free, but it feels incomplete to me if it doesn't work like -r in other places.

That is fine with me, I only added it in case rank 0 was overloaded with tasks, some could be migrated to other ranks. (We can keep it in the protocol, but remove it from the command line for now)

grondo commented:
Hm, I've found at least one dumb thing I'm doing, let me fix that up and see if the issue still occurs.

grondo commented:
Ok, I had some cut&paste code from another module to register the `cron_ctx_t` with the handle along with a free function. I noticed in backtrace that the ctx was being destroyed during handle destruction due to this free function. I removed this registration (wasn't even needed), and manually call `cron_ctx_destroy()` after the reactor returns, and this problem at least goes away.

Sorry for the noise.

garlick commented:
Ah OK, that makes sense.   I guess that is a little bit fragile then.  Sorry about that.  I'll open a bug to document.

garlick commented:
If you get rid of  `-r` you won't need to log the rank - it'll be there in the `flux exec` command that is logged.  Sorry for my contradictory comments.

grondo commented:
I removed the `-r` option but not the rank field of the cron entry at this point. I also removed `--task-history-count` option. Both these options can be set via `-o key=val` in the `flux cron (interval|event)` cmdline, e.g. `flux cron interval -o rank=1 ...` or `flux cron interval -o rank=1,task-history-count=5 ...` (but they are undocumented at this point)

garlick commented:
That clang travis failure looks like session infant mortality with very little data to go on:
```
flux-start: 2 (pid 74176) Killed
FAIL: t1003-mecho.t
```
Most likely not related to this PR!

grondo commented:
Sorry for the noise -- can't seem to get the test for cron task timeouts to work in travis... I'll move the churn to another branch for now...

grondo commented:
Just pushed a checkpoint where cron job timeouts actually work. The latest change actually changes *all* "`cmb.exec`" processes to be placed into their own process group, and `cmb.signal` sends signal to the whole process group. The issue otherwise was that cron tasks launched with `sh -c ...` would end up with the shell being killed at timeout, but not the actual processes.

In the meantime I've gone off on a tangent adding better support for hooks in the `libsubprocess` code, normalizing their usage, etc. I'm not sure if that is so useful now, so if the above temporary w/a is ok for this PR, I could park my current work, and instead work directly on finishing this up (still need docs and UI cleanup, etc.)

grondo commented:
Ok, the latest here is more uninspired work on the subprocess class.

This latest commit attempts to unify the various callback/hook points we might need for subprocess execution. The "status", "completion", and "pre-exec" hook function typedefs are unified into a `subprocess_cb_f` which takes only a `struct subprocess *` argument. Caller's context can be retrieved via `subprocess_context_get()` from a previous `subprocess_context_set()`.

The new scheme is built on an array of lists of callbacks per subprocess "event", hooks are appended to the list with `subprocess_add_hook()`, and at each hook-point, all callbacks are executed in order. This should be useful (I hope) in the future for containment setup and other various execution environment customization via plugins. Currently supported hook points are
found in `subprocess.h`:
```
/* Supported hook entry points:
 */
typedef enum {
    SUBPROCESS_PRE_EXEC,   /* In child, just before exec(2)                  */
    SUBPROCESS_POST_FORK,  /* In parent, after fork(2), before child exec(2) */
    SUBPROCESS_RUNNING,    /* In parent, after child exec(2)                 */
    SUBPROCESS_STATUS,     /* Any change in status from waitpid(2)           */
    SUBPROCESS_EXIT,       /* Subprocess has exited and been reaped          */
    SUBPROCESS_COMPLETE,   /* Subprocess has exited and all I/O complete     */
    SUBPROCESS_HOOK_COUNT
} subprocess_hook_t;
```

The IO callback should probably be moved to this same scheme, but that will require more work modifying the underlying zio stuff, or replacing it with something similar to @garlick's work in #468, which is long overdue for a review by me anyway.

I'm not sure if I've actually improved anything here, but getting rid of the seperate "status" and "completion" callbacks, each with their own `void *` args seems like a good step.

grondo commented:
And @garlick, I don't mean to be ignoring your complaints about how the subprocess code interacts with the reactor with this work. If you feel this is really headed in the wrong direction just let me know while I'm in here mucking with this stuff! I have to admit I haven't really wrapped my head around another way to wire up this kind of subprocess management in the `flux_watcher_t` style. If you have some ideas, please share.

I'm going to look seriously at your PR #468 at this point, so I think this might be a good time to fix all this weirdness.

garlick commented:
Or that could be another PR, later...

grondo commented:
Some further minor improvements:
 * Added `flux cron sync` command to control sync event behavior of loaded cron module, e.g.
   ```
   $ flux cron sync
   sync: sync disabled.
   $ flux cron sync hb
   sync: sync to event "hb" epsilon=0.015s.
   $ flux cron sync --epsilon=.1 hb
   sync: sync to event "hb" epsilon=0.100s.
   $ flux cron sync --disable
   sync: sync disabled.

   ```
 * Add `--kill` option to `flux cron delete` to kill any running task for a deleted cron entry. Without `--kill` tasks and entry stay active until task completes, at which time the entry is removed.
 
 * Add `--after` option for `flux-cron interval` to set the `after` parameter to `flux_timer_watcher`.
 
 * More tests

 * Initial version of `flux-cron(1)` manpage

grondo commented:
Some questions pertaining to finishing up the final few required tasks on this PR:

 * Should the cron module be loaded by default, and if so should `sync=hb` be set?
 * As an initial use case for `flux-cron` should a cron job to clean up old `lwj.ID` entries be added, and if so what exactly do we think it should do? Run once every 1000 jobs and clean up the oldest 100? Should it be tunable via some kvs configuration or broker attributes?

garlick commented:
The description for `sync` in the `flux-cron` usage message needs an update
```
Supported commands:
 help         Display this help or help for a command
 interval     run COMMAND once every INTERVAL
 event        run COMMAND on every event matching TOPIC
 dump         Dump values from a cron entry
 list         List registered and stopped fluxicron jobs
 delete       Delete flux cron entries
 stop         Stop a flux cron job from executing
 start        Start a stopped flux cron job
 sync         Start a stopped flux cron job
```

> Should the cron module be loaded by default

I think yes

> should sync=hb be set

I'd vote for setting it and we'll see how we like it :-)

garlick commented:
> As an initial use case for flux-cron should a cron job to clean up old lwj.ID entries be added, and if so what exactly do we think it should do? Run once every 1000 jobs and clean up the oldest 100? Should it be tunable via some kvs configuration or broker attributes?

That sounds good.  What about making a little script that takes those parameters on the command line, and install it via rc1 with default values on the script command line; then if the user wants to change from the defaults, they can `flux cron delete` and `flux cron event` with new values?

(If there was an argument for `flux cron event` similar to `flux cron interval --after`, then you could start after 1000 jobs, and then run every 100, cleaning up oldest 100...

grondo commented:
Thanks! Got the help string updated for `flux-cron sync`. I'll put 

> (If there was an argument for flux cron event similar to flux cron interval --after, then you could start after 1000 jobs, and then run every 100, cleaning up oldest 100...

Good idea!

I also thought it might be useful for the "event" cron jobs to have a parameter for "minimum time between runs" (if there's a good word for that I can't think of it right now). E.g. if you are running high throughput, you might be running 100 jobs/sec, so you might want to wake up every 100 jobs or every 15s, whichever is longer?

It seems like there should be some more flexible way to express when these jobs should run, but it is currently eluding me, so I think I'll just do the dumb thing for now.

garlick commented:
> I also thought it might be useful for the "event" cron jobs to have a parameter for "minimum time between runs"

That sounds useful, but I agree it can probably wait.  (maximum repetition rate?)

grondo commented:
For now, what about a `flux wreck purge` command? Then we don't have to worry about where to keep to the script. What I'm playing with now is something like:
```
$ flux wreck purge --help
Usage: flux-wreck purge [OPTIONS]
Purge old jobs entries from kvs
Options:
  -h, --help                 Display this message.
  -t, --target-size=N        Specify a target number of lwj entries to keep
  -R, --remove               Remove all eligible lwj entries.
  -m, --max=N                Maximum number of entries to target.
```

Unfortunately, we can't operate the purge solely from the `lwj.` directory object, because we'll have to check at least each `lwj.ID.state` entry to make sure it is "complete". Since we're doing that anyway, I also will sort the entries by completion time, so we can remove the oldest entries first.

For a local-node instance with ~4K entries, the purge analysis is taking about 1s. If this is too long there might be some optimizations we can make, or we could implement this thing as a service that is always running, in which case it could track jobs that should be purged based on lwj events with zero overhead.

garlick commented:
`flux wreck purge` sounds great.

Would it be safe to get read the lwj directory, read the lwj-active directory, and subtract?

Or go get the lowest numbered target-size lwj entries, and only check `lwj.ID.state` on those?

grondo commented:
> Would it be safe to read the lwj directory, read the lwj-active directory, and subtract?

Yes, good idea,, that should be safe for now.

> Or go get the lowest numbered target-size lwj entries, and only check lwj.ID.state on those? Possibly going back for more if too many were active?

Unfortunately the lwj ID isn't guaranteed to have any correlation to when the job completed, so I'm not sure it is safe to only process the lowest numbered IDs. lwj.4000 could have run and completed long before lwj.1 if a scheduler is active.

Unless you are saying that for now pure lwj id-based purging is good enough? Which is fine with me, but eventually we'll have to solve this problem in earnest.

If we had the lua scriptlet KVS "query" interface a tool could send a script that returned a chunk of IDs of a certain size ordered by completion time or anything else... (even a query to return complete LWJ info for a chunk of jobs instead of grabbing a key-at-a-time) That would cut down on number of round trips for this kind of workload...

garlick commented:
Oh right!  Never mind.

Yeah that lua query idea is a good one.  Maybe I should get on that!  Well not right this minute :-)

grondo commented:
Ok, I realized the script actually only needs to read 1 key per LWJ, the job's "complete-time". This is not only the key for the sort, but if the key doesn't exist for a lwj, then the job has not completed (duh.)

 This change, along with a switch from using `kvsdir` to `kvs_get` reduced the time for the test script by 1/2. I bet this could further be improved by pipelining the kvs_get requests if that is possible in the current api.

Eventually sounds like we'll have a way to request something like `lwj.*.complete-time` and that will probably reduce the time for the script by an order of magnitude.

grondo commented:
I've just pushed the current snapshot of `flux wreck purge` for feedback. It is not super quick, the script spends most of its time in the loop fetching each job's completion time. If we had a kvs_get_bulk() or something that pipelined a stream of kvs_get requests, or a way to return an rpc object for the kvs_get so responses could be handled asynchronously, that might help significantly.
```
$ flux wreck help purge
Usage: flux-wreck purge [OPTIONS]
Purge old jobs entries from kvs
Options:
  -h, --help                 Display this message.
  -v, --verbose              Increase verbosity.
  -t, --target-size=N        Specify a target number of lwj entries to keep
  -R, --remove               Remove all eligible lwj entries.
  -m, --max=N                Maximum number of entries to target.

$ flux wreck purge
purge: 2000 total job entries
$ time flux wreck purge -t 1000
purge: would remove 1000 lwj entries: 3112-4111

real	0m0.498s
user	0m0.072s
sys	0m0.021s

$ flux wreck purge -t 1000 -v  -R
purge: 0.012s: got lwj list
purge: 0.537s: gathered all completion times
purge: 0.540s: sorted on completion-time
purge: 0.854s: unlinked 1000 entries
purge: removed 1000 lwj entries: 3112-4111
purge: 0.855s: all done.

```

grondo commented:
Hm, `flux-wreck purge` isn't doing so well when run as a cron job under stress testing (running jobs as quickly as possible). The script is probably being starved out by the kvs activity of the executing jobs. In this stress test case, the "gather all job completion times" goes from about 1s to ~120s:

This session was set up with something like:
```
$ flux module load cron sync=hb
$ flux cron event --name=purge -n 500 wreck.state.complete flux wreck purge -v -t 1000 -R 
$ for i in `seq 0 4192`; do flux wreckrun -n 2 -d sleep $(echo "$RANDOM / 10000" | bc -l); done
```

And here's the log of when the cron job finally ran:

```
[1463592882.671148] cron.info[0]: cron-3[purge]: rank=0: command="flux wreck purge -v -t 1000 -R": "purge: 0.012s: got lwj list"
[1463592882.671176] cron.info[0]: cron-3[purge]: rank=0: command="flux wreck purge -v -t 1000 -R": "purge: 0.014s: Filtered 25 active jobs"
[1463592882.671189] cron.info[0]: cron-3[purge]: rank=0: command="flux wreck purge -v -t 1000 -R": "purge: 117.718s: gathered all completion times"
[1463592882.671201] cron.info[0]: cron-3[purge]: rank=0: command="flux wreck purge -v -t 1000 -R": "purge: 117.721s: sorted on completion-time"
[1463592882.671211] cron.info[0]: cron-3[purge]: rank=0: command="flux wreck purge -v -t 1000 -R": "purge: 117.906s: unlinked 1027 entries"
[1463592882.671221] cron.info[0]: cron-3[purge]: rank=0: command="flux wreck purge -v -t 1000 -R": "purge: removed 1027 lwj entries: 8824,8827,8830,8832-9852,9854-9856"
[1463592882.671232] cron.info[0]: cron-3[purge]: rank=0: command="flux wreck purge -v -t 1000 -R": "purge: 117.907s: all done."
```

We might have to rethink this approach. Single RPCs seem to return in acceptable time. I wonder if there is some easy way to return the LWJ id list in completion-sorted order...

garlick commented:
Idea: upon job completion of `lwj.N`, create `lwj-complete.H.N`as a symlink to `lwj.N`, where H is the current heartbeat epoch.  Do it under the same commit as the move from `lwj-active` to `lwj` to minimize overhead.

Then `flux wreck purge` could just walk lwj-complete in sorted order...

grondo commented:
This seems like a pretty good idea. Is the `H` there just to ensure uniqueness? Is that needed if each lwj id is guaranteed to be unique within a session? 

 I think I can try this out fairly easily...

grondo commented:
Oh nevermind about the `H` question, this obviously keeps completed jobs in sorted order. Should we bother to keep each jobid under a subdirectory of `H`? I think that is what threw me for a sec. I'm sure you have a good reason for this, but I'm clearly not thinking clearly ;-)

grondo commented:
Actually that brings up another point. In the pool I was wondering if there was any way for directory entries in the kvs to have an "mtime". The last thing that happens to a lwj dir is that its state is updated to "complete" so if we had a way to return ctime,mtime with a `kvsdir` object and return sorted order based on that -- that might be more generically helpful.

I'm sure I'm grossly oversimplifying the kvs so I apologize for that (and for wasting your time)

garlick commented:
I was thinking of fetching batches of completed jobs by epoch, thus avoiding ingesting the whole`lwj-completed` directory, but that could cut both ways -  if not very many jobs are completing per epoch then you'd have more rpcs to make.

grondo commented:
Ah, yes, you could fetch each `lwj-completed.H` in turn until you've found N entries that need to be removed. That seems like it could work well, though if the purge is working as designed then `lwj-completed.` would never grow too large. We could try it both ways and compare...

garlick commented:
> I was wondering if there was any way for directory entries in the kvs to have an "mtime"

Maybe although adding metadata to directories could make them impossible to squash together in the content store.  I'm not sure if that has a far reaching effect or not!  Have to ponder....  Obviously we would need API functions flike a `kvs_stat()`...  Hmm...

grondo commented:
A first observation while wiring this up -- it feels like it should be easier for flux handle users to access the current epoch. The "hb" event is used in enough places it almost feels like it should be directly supported by the flux handle, e.g. via a flag or `flux_enable_heartbeat()` builtin call.

As it is, a handle user has to set up a message handler, decode heartbeat messages in the handler, use `flux_event_subscribe` and remember to `unsubscribe` on exit, and also store the current epoch somewhere. It would be so much simpler if we could `flux_open (uri, FLUX_O_EPOCH)`, then access current epoch anywhere with `int flux_get_epoch (handle)`

Sorry if I'm just being lazy!!

garlick commented:
No you are right.  As we discussed just now, I added a `heartbeat-epoch` broker attribute here - feel free to cherry pick to this PR if useful: eb189e5a8e1b63c84d19fabc455ffacdc889b3b3

grondo commented:
Great thanks! I will use that. In wrexecd the job may start and complete within one heartbeat, so we'd need some way to get the current epoch anyhow. The broker attr is just a lot easier.

grondo commented:
@garlick, your suggestion above really helps, though during stress test the `flux-cron purge` still takes many seconds. However, all of the time seems to be spent doing the serial unlinks, so if we had a way to solve that it would help greatly.

Also during stress test I did some periods where the kvs was blocked for many seconds. These have not yet turned into any kind of hard hang, and I'm still able to contact other services during this time.
As requested, here's the kvs stats:
As you can see there was a commit that took ~33s
```
$ flux comms-stats kvs
{
  "obj size total (MiB)":23.172879,
  "obj size (KiB)":{
    "count":1315,
    "min":0.003000,
    "mean":18.477966,
    "stddev":54.432337,
    "max":198.943000
  },
  "#obj dirty":0,
  "#obj incomplete":1,
  "#pending commits":0,
  "#pending fences":0,
  "#watchers":0,
  "gets (sec)":{
    "count":960170,
    "min":0.000004,
    "mean":0.000068,
    "stddev":0.000727,
    "max":0.031868
  },
  "puts (sec)":{
    "count":557243,
    "min":0.000003,
    "mean":0.000022,
    "stddev":0.000174,
    "max":0.015143
  },
  "commits (sec)":{
    "count":65886,
    "min":0.000029,
    "mean":0.006885,
    "stddev":0.146005,
    "max":32.646799
  },
  "fences after sync (sec)":{
    "count":21798,
    "min":0.000168,
    "mean":0.004605,
    "stddev":0.001969,
    "max":0.050213
  },
  "commits per update":{
    "count":58134,
    "min":0.000000,
    "mean":1.133347,
    "stddev":0.461444,
    "max":15.000000
  },
  "#no-op stores":302659,
  "#faults":334,
  "store revision":73477
}

```

garlick commented:
Yikes!  The 32s commit is pretty disturbing.  I wonder if that was a case where rank 0 had to fault in something involved in a commit?  The commit code is all serial.   

I'm not sure this has to be solved right now but if you're curious, you could try tuning the content cache purge attributes on rank 0, e.g.
```
flux setattr content-purge-large-entry 8192  # default was 256
```
From flux-broker-attributes(7):
>  *content-purge-large-entry*
           When the cache size footprint needs to be reduced, first consider
           purging entries of this size or greater.

>  *content-purge-old-entry*
           When the cache size footprint needs to be reduced, only consider
           purging entries that are older than this number of heartbeats.

>  *content-purge-target-entries*
 content cache purge attributes on rank 0, e.g.
```
flux setattr content-purge-large-entry 8192  # default was 256
```
From flux-broker-attributes(7):
>  *content-purge-large-entry*
           When the cache size footprint needs to be reduced, first consider
           purging entries of this size or greater.

>  *content-purge-old-entry*
           When the cache size footprint needs to be reduced, only consider
           purging entries that are older than this number of heartbeats.

>  *content-purge-target-entries

>  *content-purge-large-entry*
           When the cache size footprint needs to be reduced, first consider
           purging entries of this size or greater.

>  *content-purge-old-entry*
           When the cache size footprint needs to be reduced, only consider
           purging entries that are older than this number of heartbeats.

>  *content-purge-target-entries*
 content cache purge attributes on rank 0, e.g.
```
flux setattr content-purge-large-entry 8192  # default was 256
```
From flux-broker-attributes(7):
>  *content-purge-large-entry*
           When the cache size footprint needs to be reduced, first consider
           purging entries of this size or greater.

>  *content-purge-old-entry*
           When the cache size footprint needs to be reduced, only consider
           purging entries that are older than this number of heartbeats.

>  *content-purge-target-entries*
           If possible, the cache size purged periodically so that the total
           number of entries stays at or below this value.

garlick commented:
On second thought, this seems likely to be a time sink and is ancillary to this PR.  I'd be fine getting this in now, as is, and dealing with KVS woes later on when we do the "treeobj" work.

I will take a quick look at `kvs_unlink()` to see if there is a quick fix there though.

garlick commented:
I don't think I can do anything for `kvs_unlink()` right now.  Eventually we will collapse the RTT's by having the commit object that is currently built in the kvs module, built in the handle and sent to the module when `kvs_commit()` is called.  `kvs_unlink()` ops have to be kept in order with the other ops appended to the commit object, so to get any relief now with that approach would mean tackling the whole problem I think.

The slightly tricky part in tackling the whole problem is that larger objects should be pushed into the content cache with an async rpc as the commit object is built, and responses to those have to be received before the commit (with refs to the objects) is sent.   I think to do that with good interfaces and internal structure is a week or two, not an afternoon, unfortunately.

So I'd vote to press on for now.

grondo commented:
Ok, sounds reasonable.

I wired up my build directory with the following in `rc1`:
```
flux module load -r 0 cron sync=hb
flux cron event -n 250 wreck.state.complete flux wreck purge -v -R -t 1024
```

Then ran the soak test with 8K jobs and 16K jobs, and things are looking pretty good. In the soak test at least the kvs rtt issue doesn't seem to be a problem. One other difference is that I ran on the soak tests on a compute node of hype, whereas my other testing was on the login node.

I'm also running another 8K job soak test without the lwj purge, but that is predictably taking longer...

![soak-9545](https://cloud.githubusercontent.com/assets/741970/15407559/579329c6-1dc0-11e6-8eb2-ff7590759c64.png)

![soak-74490](https://cloud.githubusercontent.com/assets/741970/15409625/791517e4-1dca-11e6-840d-35159e811282.png)

garlick commented:
Victory!

grondo commented:
For completeness, here's the same 8K soak test without the cron job. It actually isn't too bad but you can see the low slope of the runtimes is still there:

![soak-180358](https://cloud.githubusercontent.com/assets/741970/15410374/38edf484-1dce-11e6-99cd-d28b03201479.png)

grondo commented:
OK, I've just pushed the changes discussed above and used in the soak test results posted above for any comments.

grondo commented:
Ok, I couldn't help but add `--after` and `--min-interval` support to `flux-cron event` this morning. 

As part of this, the "stats" for a cron entry were more fully fleshed out, including counters for total number of times the entry has been run, number of successes (non-zero exit) and failures (anything else), creation time, last time executed, and last start time of the entry. The entry object was moved to a cron-wide header so that the various cron types can use this information for modifying their behavior.

grondo commented:
One of the new tests is failing intermittently in Travis, and I'm currently debugging on another branch.

grondo commented:
Ok, I've just pushed a rebased & squashed version of this PR, with all the intermediate work removed.

Still todo is to add `flux module load cron` to the default `rc1` script -- if we think we want that.

Also, someone should verify  that the interfaces presented in `flux-cron(1)` are acceptable.

grondo commented:
Also, as @garlick just commented, perhaps `flux wreck purge` should optionally stream deleted job information to a file? Or, jobids and content-store hashes? If the second I need a scheme to get those from the kvs as I delete the jobs -- this will probably slow down the purge a little bit more...

garlick commented:
> perhaps flux wreck purge should optionally stream deleted job information to a file

I think that feature need not hold up this PR.  We can always add it later.

garlick commented:
This is all looking good.  I meant to put it through its paces this afternoon but it turns out I need to leave for home right now.  Will try again tomorrow or if you think we've already worked this over enough I'm also fine with pressing the button.

garlick commented:
Poking at this a bit.  In flux-cron(1) you might mention in the DESCRIPTION that the exec service bypasses the scheduler and provide a reference to flux-exec(1).  Also you might mention in there that stdio and stderr are logged and can be accessed with flux-dmesg(1).

Should a nonzero exit for a cron job be logged and/or cause the job to stop?

grondo commented:
Ok, great comments. I'll work on the man page a little more.

Nonzero exit should probably be logged, though I've noticed log messages don't survive very long in `flux dmesg` ring buffer in general, so they are also accumulated in `stats.failure` field of the entry stats structure, available for now with `flux cron dump`.

I went back and forth on whether to stop jobs by default on failure. I thought the least surprising choice was to continue to attempt to run the job and accumulate the number of successes and failures -- but perhaps this should be an option set via `-o OPTIONS`?

garlick commented:
> I've noticed log messages don't survive very long in flux dmesg ring buffer in general

LOG_ERR and above should go to stderr also.

If the failure was more visible, would you change your mind about stopping the job on failure?  I lean that way somewhat.

grondo commented:
> LOG_ERR and above should go to stderr also.
This will scroll off the screen even more quickly than the dmesg ring buffer is exhausted.

> If the failure was more visible, would you change your mind about stopping the job on failure? I lean that way somewhat.

Not really, but I'm willing to go whichever way you want. I did note that none of the existing cron-type systems seem to stop cron jobs on failure (though I may have misread). Airbnb's cronos seemed to have the nicest approach where they accumulated failures and successes by default. On checking the cron job status if you see a large number of failures you can examine the script.

Take for example the specific use case for this PR. Do you really want to stop the `flux wreck purge` cron job by default on a single failure? If it was an intermittent failure, I would want the purge to keep retrying instead of stopping, so that the lwj dir was kept at the expected level.

However, I'm perfectly fine making this a tunable. It could be a `stop-on-failure = N` setting, where we stop on the `Nth` failure, 1 by default, -1 for never stop?

garlick commented:
You make some good points!  I'm sold on the current behavior.  Might be worth mentioning in the man page though.

grondo commented:
Ok, I added `stop-on-failure` support, and pretty much rewrote the manpage, but I hope to didn't go from a-bit-too-terse to annoyingly-verbose. Nonzero exit codes are also now logged at `LOG_ERR`

garlick commented:
Man page looks good.

I think `stop-on-failure` was typoed to `stop-on-exec` a couple of times in there though.

grondo commented:
Thanks for catching the errors in the man page, I'll fix those up this morning, and I'm also working on test for `stop-on-failure`

Hm, strange error in Travis for one of the builds, I'll try a restart.
```
[1464148127.980060] broker.err[0]: rc1: flux-module: flux_open: No such file or directory
[1464148127.980504] broker.err[0]: Run level 1 Exited with non-zero status (rc=1)
[1464148128.108393] broker.err[0]: rc3: flux-module: cmb.rmmod[0] wrexec: No such file or directory
[1464148128.131436] broker.err[0]: rc3: flux-module: cmb.rmmod[0] job: No such file or directory
[1464148128.155920] broker.err[0]: rc3: flux-module: cmb.rmmod[0] mecho: No such file or directory
[1464148128.179890] broker.err[0]: rc3: flux-module: cmb.rmmod[0] resource-hwloc: No such file or directory
[1464148128.204612] broker.err[0]: rc3: flux-module: cmb.rmmod[0] kvs: No such file or directory
[1464148128.228490] broker.err[0]: rc3: flux-module: cmb.rmmod[0] barrier: No such file or directory
[1464148128.255138] broker.err[0]: rc3: flux-module: cmb.rmmod[0] content-sqlite: No such file or directory
FAIL: lua/t1004-statwatcher.t
```

garlick commented:
Hmm, I wonder if maybe the retry logic in the local connector wasn't persistent enough for a slow connector-local module startup?  I will look at that in a bit to see if anything can be improved there.

grondo commented:
Just fixed manpage typos and pushed the tests for `stop-on-failure` option.

garlick commented:
This would indicate that the `connector-local` unix domain socket socket has not yet been created:
```
[1464148127.980060] broker.err[0]: rc1: flux-module: flux_open: No such file or directory
```
The `connector-local` module is hardwired to be loaded before the rc1 script executes, however the broker doesn't wait for the module's thread to complete its initialization and start the reactor like the external module loader does.  So despite the fact that the handle code includes some retry logic, I think there is still a race here.  I'll open an issue.

grondo commented:
Strange that travis didn't automatically fire off with the latest changes. I'll squash things up and force a push to see if that does anything. Then I think we're getting close to ready for this particular version of the PR.

grondo commented:
I've just force pushed an update with newer commits squashed back together.

Also, the cron module is now loaded by default in `rc1` with `sync=hb` set.

grondo commented:
Maybe give this one more test since I added `flux module load -r 0 cron sync=hb` to `rc1`, but then I think this version may be ready for merging...

garlick commented:
Couple of small nits

Would it make sense to move the CRON COMMANDS to the top of the man page for easier reference?

Maybe something amiss with argument error handling here?
```
$ flux cron stop
lua: ...me/garlick/proj/flux-core/build/../src/cmd/flux-cron:396: attempt to call method 'usage' (a string value)
stack traceback:
	...me/garlick/proj/flux-core/build/../src/cmd/flux-cron:396: in function 'handler'
	...core/build/../src/bindings/lua/flux/Subcommander.lua:146: in function <...core/build/../src/bindings/lua/flux/Subcommander.lua:127>
	(tail call): ?
	...me/garlick/proj/flux-core/build/../src/cmd/flux-cron:463: in main chunk
	[C]: ?
```

Would it be a pain to decode the wait status here?
```
1464200233.061718] cron.err[0]: cron-1: "/bin/false": exited with status 0x0100
```

grondo commented:
Good points. I'm up for *any* comments on the man page I didn't really like the end result.

I can fix the other issues you've discovered. Thanks!

grondo commented:
Is this a bit better?
```
[1464209426.095469] cron.err[0]: cron-3: "false": Failed: Exit 1
```

```
[1464209384.097531] cron.err[0]: cron-2: "kill -11 $$": Failed: Segmentation fault
```

garlick commented:
Yes!

grondo commented:
Ok, force pushed a new version with comments by @garlick addressed (I think).

garlick commented:
Great work!  Merging.
